internal/manifest/manifest.go

```diff
@@ -140,87 +142,157 @@ func validateHexString(value HexString, expectedNumBytes int) error {
 func (m *Manifest) Validate() error {
+	var errs []error
 	for policyHash, policy := range m.Policies {
 		if err := policy.Validate(policyHash); err != nil {
-			return fmt.Errorf("validating policy %s: %w", policyHash, err)
+			errs = append(errs, newValidationError(fmt.Sprintf("Policies[%q]", policyHash), err))
 		}
 	}
 
 	if err := m.ReferenceValues.Validate(); err != nil {
-		return fmt.Errorf("validating reference values: %w", err)
+		errs = append(errs, err)
 	}
 
-	for _, keyDigest := range m.WorkloadOwnerKeyDigests {
+	for i, keyDigest := range m.WorkloadOwnerKeyDigests {
 		if _, err := keyDigest.Bytes(); err != nil {
-			return fmt.Errorf("decoding key digest %s: %w", keyDigest, err)
+			errs = append(errs, newValidationError(fmt.Sprintf("WorkloadOwnerKeyDigests[%d]", i), err))
 		} else if len(keyDigest) != hex.EncodedLen(sha256.Size) {
-			return fmt.Errorf("workload owner key digest %s has invalid length: %d (expected %d)", keyDigest, len(keyDigest), hex.EncodedLen(sha256.Size))
+			errs = append(errs, newValidationError(fmt.Sprintf("WorkloadOwnerKeyDigests[%d]", i), fmt.Errorf("invalid length: %d (expected %d)", len(keyDigest), hex.EncodedLen(sha256.Size))))
 		}
 	}
 
-	for _, key := range m.SeedshareOwnerPubKeys {
+	for i, key := range m.SeedshareOwnerPubKeys {
 		if _, err := ParseSeedShareOwnerKey(key); err != nil {
-			return fmt.Errorf("invalid seed share owner public key %s: %w", key, err)
+			errs = append(errs, newValidationError(fmt.Sprintf("SeedshareOwnerPubKeys[%d]", i), err))
+		}
+	}
+	return errors.Join(errs...)
+}
+
+// validationError contains a JSON path and a list of errors.
+// Nested validation errors are printed on newlines with the full path.
+type validationError struct {
+	path string
+	errs []error
+}
+
+func newValidationError(path string, errs ...error) error {
+	e := &validationError{
+		path: path,
+		errs: make([]error, 0, len(errs)),
+	}
+	for _, err := range errs {
+		if err != nil {
+			e.errs = append(e.errs, flatten(err)...)
+		}
+	}
+	if len(e.errs) == 0 {
+		return nil
+	}
+	return e
+}
+
+func (e *validationError) Error() string {
+	return e.formatError(e.path)
+}
+
+func (e *validationError) Unwrap() []error {
+	return e.errs
+}
+
+func (e *validationError) formatError(path string) string {
+	var sb strings.Builder
+	for i, err := range e.errs {
+		var ve *validationError
+		if errors.As(err, &ve) {
+			sb.WriteString(ve.formatError(path + "." + ve.path))
+		} else {
+			sb.WriteString(path)
+			sb.WriteString(": ")
+			sb.WriteString(err.Error())
+		}
+		if i < len(e.errs)-1 {
+			sb.WriteString("\n")
+		}
+	}
+	return sb.String()
+}
+
+func flatten(err error) (errs []error) {
+	if ve, ok := err.(*validationError); ok { //nolint:errorlint // check for exact type
+		return []error{ve}
+	}
+	if wrapped, ok := err.(interface{ Unwrap() []error }); ok {
+		for _, err := range wrapped.Unwrap() {
+			errs = append(errs, flatten(err)...)
 		}
+		return errs
 	}
-	return nil
+	return []error{err}
```

Comment:
For completeness you could cast into `Unwrap() error` and recurse into that one, too.

Reply:
As discussed, not needed.